### PR TITLE
ci: refactor and bump actions

### PR DESCRIPTION
**Pull request by ffranr**
This PR refactors the CI pipeline and also bumps action version numbers.

It also introduces a new action which sets up the go environment.

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -72,15 +72,46 @@ jobs:
...
+      - name: setup go ${{ env.GO_VERSION }}
+        uses: actions/setup-go@v2
+        with:
+          go-version: '${{ env.GO_VERSION }}'
+
       - name: run lint
         run: make lint
 
+
+   ########################
```
Comment (bhandras) [inline .github/workflows/main.yaml:97]: nit: formatting is a little bit off

**Review by bhandras — APPROVED** [review]

**Review by positiveblue — APPROVED** [review]
In this PR there is still a `Format, compilation and lint check` step in the CI (not executed) marked as required

Is that a github glitch?

**Comment by ffranr** [comment]
> In this PR there is still a `Format, compilation and lint check` step in the CI (not executed) marked as required
> 
> Is that a github glitch?

I think this is related to branch protection rules, see [here](https://github.com/orgs/community/discussions/26698#discussioncomment-3252954). Protection rules probably require specific jobs to pass. They refer to these jobs by name. The rules will need to be updated before merging this PR.

I'm not sure if @Roasbeef is available today. Is there anyone else that can modify these rules?

**Review by bhandras — APPROVED** [review]
LGTM :+1:

**Pull request merged**